syntax = "proto3";

option go_package = "github.com/tweag/trustix/proto";

package trustix;

// TrustixKV - Access a remote Key/Value store (remote log)
service TrustixKV {
  // SubmitMapping - Submit an input/output mapping to the local log
  rpc Get (KVRequest) returns (KVResponse) {}
}

message KVRequest {
  bytes Bucket = 1;
  bytes Key = 2;
}

message KVResponse {
  bytes Value = 1;
}

// TrustixCombinedRPC
service TrustixCombinedRPC {
  // Get map[LogName]OutputHash
  rpc HashMap (HashRequest) returns (HashMapResponse) {}
  // Compare(inputHash)
  rpc Decide (HashRequest) returns (DecisionResponse) {}

}

message HashRequest {
  bytes InputHash = 1;
}

message HashMapResponse {
  map<string, bytes> hashes = 1;
}

message OutputHashResponse {
  string LogName = 1;
  bytes OutputHash = 2;
}

message OutputHashDecision {
  repeated string LogNames = 1;
  bytes OutputHash = 2;
  int32 Confidence = 3;
}

message DecisionResponse {
  OutputHashDecision Decision = 1;

  // Non-matches (hash mismatch)
  repeated OutputHashResponse Mismatches = 2;

  // Full misses (missing log entry)
  repeated string Misses = 3;
}